0001-Add-the-camera-sensor-gc5035-to-support-ADL-M.patch

```diff
@@ -0,0 +1,2471 @@
++		ret = -EINVAL;
++		goto out;
++	}
++
++out:
++	v4l2_fwnode_endpoint_free(&bus_cfg);
++	fwnode_handle_put(ep);
++	return ret;
++}
++
++static int gc5035_probe(struct i2c_client *client)
++{
++	struct device *dev = &client->dev;
++	struct gc5035 *gc5035;
++	struct v4l2_subdev *sd;
++	int ret, i;
++	u32 freq = 192000000UL;
++
++	gc5035 = devm_kzalloc(dev, sizeof(*gc5035), GFP_KERNEL);
++	if (!gc5035)
++		return -ENOMEM;
++
++	gc5035->client = client;
++	gc5035_init_power_ctrl(gc5035);
++	gc5035_set_power(gc5035, 1);
++	
++	gc5035->cur_mode = &gc5035_modes[0];
++
++	ret = clk_set_rate(gc5035->mclk, freq);
++	if (ret < 0)
++		return dev_err_probe(dev, ret, "Failed to set mclk rate\n");
++	gc5035->mclk_rate = clk_get_rate(gc5035->mclk);
++	if (gc5035->mclk_rate != freq)
++		dev_warn(dev, "mclk rate set to %lu instead of requested %u\n",
++			 gc5035->mclk_rate, freq);
++	gc5035->iovdd_supply = devm_regulator_get(dev, "iovdd");
++	if (IS_ERR(gc5035->iovdd_supply))
++		return dev_err_probe(dev, PTR_ERR(gc5035->iovdd_supply),
++				     "Failed to get iovdd regulator\n");
++
++	for (i = 0; i < ARRAY_SIZE(gc5035_supplies); i++)
++		gc5035->supplies[i].supply = gc5035_supplies[i];
++	ret = devm_regulator_bulk_get(&gc5035->client->dev,
++				       ARRAY_SIZE(gc5035_supplies),
++				       gc5035->supplies);
++	if (ret)
++		return dev_err_probe(dev, ret, "Failed to get regulators\n");
++    
++	mutex_init(&gc5035->mutex);
++	sd = &gc5035->subdev;
++	v4l2_i2c_subdev_init(sd, client, &gc5035_subdev_ops);
++	ret = gc5035_initialize_controls(gc5035);
++	if (ret) {
++		dev_err_probe(dev, ret, "Failed to initialize controls\n");
++		goto err_destroy_mutex;
++	}
++	ret = gc5035_runtime_resume(dev);
++	if (ret) {
++		dev_err_probe(dev, ret, "Failed to power on\n");
++		goto err_free_handler;
++	}
++	ret = gc5035_check_sensor_id(gc5035, client);
++	if (ret) {
++		dev_err_probe(dev, ret, "Sensor ID check failed\n");
++		goto err_power_off;
++	}
++
++	sd->flags |= V4L2_SUBDEV_FL_HAS_DEVNODE;
++	sd->entity.ops = &gc5035_subdev_entity_ops;
++	sd->entity.function = MEDIA_ENT_F_CAM_SENSOR;
++	gc5035->pad.flags = MEDIA_PAD_FL_SOURCE;
++	ret = media_entity_pads_init(&sd->entity, 1, &gc5035->pad);
++	if (ret < 0) {
++		dev_err_probe(dev, ret, "Failed to initialize pads\n");
++		goto err_power_off;
++	}
++	ret = v4l2_async_register_subdev_sensor(sd);
++	if (ret) {
++		dev_err_probe(dev, ret, "v4l2 async register subdev failed\n");
++		goto err_clean_entity;
++	}
++
++	pm_runtime_set_active(dev);
++	pm_runtime_enable(dev);
++	pm_runtime_idle(dev);
++	gc5035_set_power(gc5035, 0);
++
++	return 0;
++
++err_clean_entity:
++	media_entity_cleanup(&sd->entity);
++err_power_off:
++	gc5035_runtime_suspend(dev);
++err_free_handler:
++	v4l2_ctrl_handler_free(&gc5035->ctrl_handler);
++	gc5035_set_power(gc5035, 0);
++err_destroy_mutex:
++	mutex_destroy(&gc5035->mutex);
++
++	return ret;
++}
++
++static void gc5035_remove(struct i2c_client *client)
```

Comment:
Could you add supporting for 5.15+ kernels?
```
#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 1, 0)
static int gc5035_remove(struct i2c_client *client)
#else
static void gc5035_remove(struct i2c_client *client)
#endif
{
...
#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 1, 0)
	return 0;
#endif
}
```

Also need to include:
```#include <linux/version.h>```

Reply:
Thanks Hao review, update done.